### ISSUE TITLE
[feature request] Improve dataset mounting and documentation for mounting 

### ISSUE THREAD

**Issue by d-uzlov** [openzfsonwindows/openzfs#245]
### Describe the feature would like to see added to OpenZFS

1. Documentation on different ways datasets can be mounted (or not mounted).

---

2. Ability to set arbitrary mount points for datasets, as folders (reparse points).

- Maybe the driver could mount dataset as `C:/<mountpoint-value>` when `mountpoint` is set but `driveletter` explicitly is disabled?
- - Datasets will need to inherit `driveletter` from parent for this to work properly
- - It would maybe break things for already existing datasets? They don't currently inherit `driveletter`, do they?
- Or maybe it should be combined with the part `3.` of this feature request

---

3. Ability to set several mount points per dataset.

- If Linux users are expected to use symlinks and/or `mount --rbind` for this, I think it would still be good to mention alternative ways for Windows in the documentation
- Even if ZFS doesn't have proper settings for this, I think it would be good to separate drive letter from filesystem mount points (reparse point)
- - For example:
- - `tank` is mounted as `Q:/` (using `driveletter`)
- - `tank/data` is mounted as `Q:/data` (using `driveletter`)
- - And at the same time:
- - `tank` is mounted as `C:/zfs-root/tank` (using `mountpoint`)
- - `tank/data` is mounted as `C:/zfs-root/custom-data-mount` (using `mountpoint`)
- I think it would be OK force mounting on the `C:` drive, to avoid incompatible `mountpoint` values between Unix and Windows.
- Though, then driver would probably need to have some way to disable double-mounting every dataset when user doesn't want this, which additionally complicates things, so, maybe this idea needs additional refinement

It would be ideal to have the ability to use standard Windows tools, like Disk Management util, to change mount points, but I guess if this isn't done already, it should be extremely hard or impossible to integrate ZFS into it.

---

### Additional context

Currently documentation is pretty base-bones IMO:

https://github.com/openzfsonwindows/openzfs/blob/1fd5a8bc21d847e6fa0ba33901836b55467b3a79/module/os/windows/README.md?plain=1#L286-L295

What I got from my experiments:
- The only way to change mount options is through `zfs set`
- - ZFS datasets don't show in the Disk Management util, which usually can change mount point(s) for the drive
- - Yeah, I guess the reason is obvious, but still
- `zfs set driveletter= pool` causes error when mounting
- `zfs set driveletter=- pool` is the proper (but undocumented) way to disable separate drive letter for a dataset
- `zfs inherit driveletter pool` doesn't work
- Datasets are mounted regardless of the `driveletter` value, in some way or another
- The proper way to avoid mounting a dataset is to set `zfs set mountpoint=none pool`
- - But `mountpoint` doesn't affect anything else
- - It doesn't change where the dataset will be mounted
- If I want to re-enable mounting, I need to set `mountpoint` to some arbitrary location, which doesn't matter
- Mounted dataset will always use a ZFS-specific drive letter
- - If root dataset is mounted at all, it will always get its own letter
- - Other datasets will either get its own letter or be mounted inside parent dataset
- - ... Except when I broke something in my experiments, `tank/data/local` dataset somehow got mounted as `C:\local`
- - - Shoutout to https://github.com/openzfsonwindows/openzfs/issues/244
- It seems impossible to get more than one mount point for a dataset

I'm probably still missing some quirks.

I think it would make the project more friendly for new users if there was a separate section in the documentation about mounting, which described all of this and had more examples.

The feature request for improving mount options would be nice to have but not critical at all.
But since current documentation mentions mounting under `Design issues that need addressing` I thought I could suggest something.

**Comment by lundman** [comment]
I don't disagree with anything here. :) At the moment I am focused on getting it working, so documentation and nice features are further down the list. But all these things are up for grabs if people want to help, fork, make changes, create PR.

The way it should work is that `driveletter` is set to a special `poolonly` value by default. Which means it is `driveletter=on` for the pool, and `driveletter=off` for any lower datasets. That way if you just import a default foreign pool, consisting of:
```
/pool/
/pool/datasetA/
/pool/datasetA/datasetB/
```

the default on Windows will be to assign a drive letter (let's say E:) for the pool:
```
E:\
E:\datasetA
E:\datasetA\datasetB
```

this "works" right now, at least, for all my tests. But there is that #238 which is probably why everyone is setting `driveletter` at the moment, to work around it. I want to fix this bug, like, I have today free to do so, but I need to replicate the issue to understand what goes wrong.

`driveletter` does not inherit for obvious reasons - If you set `driveletter=G` then obviously all lower datasets can not also be `G:`. Potentially setting `driveletter=on` (or off) could inherit, but I don't think that would add anything really. The default is enough for most people (when it works).

Changing `driveletter` (like `mountpoint`) is supposed to re-mount automatically like upstream, but `unmount` is such an iffy operation at the moment, I have disabled the automatic part.

> zfs set driveletter=- pool

Neat, had never thought to unset - so this should be fixed.

The multiple mounting isn't something I had considered, but now that you point it out, you totally can do that in Windows, since volumes aren't "mounted", so reparse_points can be anywhere. It's an interesting idea, and syntax for such a thing would need to be thought about. 


Ultimately, I am hoping to attract a Windows UI developer (since I am banned from doing UI, by edicts of the pope) at least initially something simple like ZetaWatch: https://github.com/cbreak-black/ZetaWatch

Finally, I suspect the mounting we do with MountMgr, is not-entirely-correct. As you point out, they don't show up in DiskManager and similar. They are in MountMgr's Registry, but beyond that they appear both there, and not there. There is very little MountMgr information out there, so I am not sure how to debug that. Just like unmounting is the "riskiest" thing at the moment, you are better off rebooting. These two probably also needs an experience Windows developer to weigh in on.

**Comment by d-uzlov** [comment]
Yeah, I understand that this is a low-priority stuff for the moment, I just wanted to write it down while my thought on this are still fresh, so I don't forget anything.
And maybe for re-reading this in the future and thinking about better ways to manage this.

Maybe I'll try to create a proper PR for documentation about current state of the implementation if I have time and energy in the next few days.

> But there is that #238 which is probably why everyone is setting `driveletter` at the moment, to work around it.

Well, I instead very quickly got to https://github.com/openzfsonwindows/openzfs/issues/244, after which I decided that I don't want to mess with nested mount points :-P

But also I treat drive letters as a very convenient shortcuts which are available everywhere.
The tree structure is superior in many ways, but when working with UI of some program, for me it's just much easier to an item at the root of the filesystem instead of specifying the whole path. There are alternatives, like shortcuts, but they aren't integrated into every program, unlike drive letters.
So this is why for my use-case I would like to mount the dataset tree as-is, fully (and maybe as part of disk C:, without a separate drive letter), plus add additional drive letters to few selected datasets.

I agree that current system is good enough, and my habits probably don't represent most other users.
